sync-core/src/main/java/com/cloudant/sync/query/QuerySqlTranslator.java

```diff
@@ -104,10 +104,23 @@ public static QueryNode translateQuery(Map<String, Object> query,
         TranslatorState state = new TranslatorState();
         QueryNode node = translateQuery(query, indexes, state);
 
-        // If we haven't used a single index or an OR clause is missing an index,
-        // we need to return a query which returns every document, so the posthoc
-        // matcher can run over every document to manually carry out the query.
-        if (!state.atLeastOneIndexUsed || state.atLeastOneORIndexMissing) {
+        if (state.textIndexMissing) {
+            String msg = "No text index defined, cannot execute query containing a text search.";
+            logger.log(Level.SEVERE, msg);
+            return null;
+        } else if (state.textIndexRequired && state.atLeastOneIndexMissing) {
+            String msg = String.format("Query %s contains a text search but is missing \"json\"" +
+                                       " index(es).  All indexes must exist in order to execute a" +
+                                       " query containing a text search.  Create all necessary" +
+                                       " indexes for the query and re-execute.",
+                                       query.toString());
+            logger.log(Level.SEVERE, msg);
```

Comment:
Added FB case 46482 - https://cloudant.fogbugz.com/f/cases/46482 to revisit this restriction in the future.